### PR TITLE
Writing flow: Backspace at beginning of first paragraph block prevents block from being deleted

### LINKED ISSUE #56303: Backspace at the beginning of document deletes current block
### Description

In a block editor with some text content in it, some paragraph blocks, placing the cursor at the start and pressing BACKSPACE will delete the 1st paragraph. 

I expect in this situation either one of:

- nothing happens
- cursor moves to title

### Step-by-step reproduction instructions

0. Be on Gutenberg trunk
1. Make a post
2. Add some paragraphs
3. Place the cursor at the start of the document content
4. Press BACKSPACE
5. See bug

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/107534/9619ac6e-6591-4e37-92f1-7f39dea5a9a7

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes #56303
Related to #55134

## What?
This PR fixes an issue where pressing the backspace key at the beginning of a paragraph block with content causes the block to be deleted.

### Before 

https://github.com/WordPress/gutenberg/assets/54422211/7f0ccf27-d486-4e60-9437-8462be6a1d63

### After

https://github.com/WordPress/gutenberg/assets/54422211/bdae7725-f403-4440-b98f-e61db4606b3f

## Why?

I think the current implementation deletes a block when you press the backspace key at the beginning of the first block. This is a problem if the paragraph or heading already has content.

## How?
I updated the logic to have the following behavior when the BACKSPACE key is pressed in the first block. I think this is expected behavior.

- If it is the default block, i.e. a paragraph block, do nothing
- If it is not a default block, convert it to the default block. This means that headings are converted to paragraph blocks while preserving their content.

## Testing Instructions

- Insert a Paragraph block at the beginning of the content and enter the text.
- Press the backspace at the beginning of the Paragraph block.
  - trunk: The block will be removed and you should now have an empty paragraph.
  - This PR: Nothing should happen.
- Insert a heading block at the beginning of the content and enter the text.
- Press the backspace key at the beginning of the Heading block.
  - trunk: The block will be removed and you should now have an empty paragraph.
  - This PR: Should be transformed to a paragraph block.

### CODE REVIEW

File: packages/block-editor/src/components/block-list/block.js [hunk 1]
```diff
@@ -451,8 +451,17 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, registry ) => {
 					}
 
 					moveFirstItemUp( rootClientId );
-				} else {
-					removeBlock( clientId );
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block.js:463]: Should we maybe just not do anything? I'm not sure if pressing Backspace at the start of a heading should convert it to a paragraph? Are there any editors that do this?

Comment (t-hamano) [reply packages/block-editor/src/components/block-list/block.js:463]: Google Docs and MS Word don't seem to transform a heading into a paragraph. Regarding Notion, it seems to transform a heading into a paragraph.

https://github.com/WordPress/gutenberg/assets/54422211/1c38ba09-6f18-4abe-913a-85e754ed2021

Personally, I like the current behavior of the Backspace key. If the backspace key does nothing, we will need to transform the existing heading into a paragraph or delete it in order to start a new article again from a paragraph. This seems a little cumbersome to me.

File: packages/block-editor/src/components/block-list/block.js [hunk 2]
```diff
@@ -451,8 +451,17 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, registry ) => {
...
+				} else if (
+					getBlockName( clientId ) !== getDefaultBlockName()
+				) {
+					const replacement = switchToBlockType(
+						getBlock( clientId ),
+						getDefaultBlockName()
+					);
+					if ( replacement && replacement.length ) {
+						insertBlocks( replacement, getBlockIndex( clientId ) );
+						removeBlock( clientId, false );
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/block.js]: Can you not use the replace blocks action?

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block.js]: If not, this needs to be batched.

Comment (t-hamano) [reply packages/block-editor/src/components/block-list/block.js]: > Can you not use the replace blocks action?

That's right. Updated by [7648597](https://github.com/WordPress/gutenberg/pull/56329/commits/76485973fd39176ffd8a0b27c9d0ecc64e837c6c).

**Review by ellatrix — APPROVED** [review]

**Pull request merged**